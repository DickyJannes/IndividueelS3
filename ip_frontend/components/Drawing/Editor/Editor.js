import React, { useState } from "react";
import { View, Text, TextInput, Button, Switch} from "react-native";
import ColorPalette from "react-native-color-palette";

import DrawingPanel from "../DrawingPanel/DrawingPanel";
import Grid from "../GridOverlay/Grid/Grid";
import editorStyles from "./editorStyle";

export default function Editor() {
	const [panelWidth, setPanelWidth] = useState(8);
	const [panelHeight, setPanelHeight] = useState(8);
	const [hideOptions, setHideOptions] = useState(false);
	const [hideDrawingPanel, setHideDrawingPanel] = useState(true);
	const [hideWidthHeightError, setHideWidthHeightError] = useState(false);
	const [buttonText, setButtonText] = useState("start drawing");
	const [selectedColor, setColor] = useState("#FFFFFF");
	const [newBackground, setBackground] = useState("#C0C1BB");

	const [isEnabled, setIsEnabled] = useState(false);
	const [hideGrid, setHideGrid] = useState(false);

	function toggleSwitch() {
		setIsEnabled(previousState => !previousState);
		setHideGrid(!hideGrid);

	} 

	function initializeDrawingPanel() {
		if(panelHeight <= 4 || panelWidth <= 4 || panelHeight >= 19 || panelWidth >= 19)
		{
			setHideWidthHeightError(true);
			return;
		}
		setHideWidthHeightError(false);
		setHideOptions(!hideOptions);
		setHideDrawingPanel(!hideDrawingPanel);

		if(isEnabled){
			setHideGrid(!hideGrid);
		}

		buttonText === "start drawing"
			? setButtonText("reset")
			: setButtonText("start drawing");
	}

	function changeColor(color) {
		setColor(color); 
	}

	function changeBackgroundColor(color) {
		setBackground(color); 
	}

  
	return (
		
		<View style={editorStyles.editor}>
			<Text style={editorStyles.h1}>Pixel Editor</Text>
			{hideDrawingPanel && <Text style={editorStyles.h2}>Enter Panel Dimensions</Text>}
			{hideWidthHeightError && <Text style={editorStyles.errorMessage}>Height / width must be between 5 and 18 pixels</Text> }
			{hideDrawingPanel && (
				<View style={{alignItems: "center", justifyContent: "center"}}>
					<View style={editorStyles.options}>
						<View style={editorStyles.option}>
							<TextInput
								testID="inputWidth"
								style={editorStyles.panelInput}
								keyboardType='numeric'
								value={panelWidth}
								onChange={(e) => { setPanelWidth(e.target.value);} 
								} 
							/>
							<Text>Width</Text>
						</View>
						<View style={editorStyles.option}>
							<TextInput
								style={editorStyles.panelInput}
								keyboardType='numeric'
								value={panelHeight}
								onChange={(e) => { setPanelHeight(e.target.value);} 
								} 
							/>
							<Text>Height</Text>
						</View>
					</View>
					<ColorPalette
						onChange={changeBackgroundColor}
						defaultColor={newBackground}
						colors={["#C0C1BB", "#000000", "#FFFFFF", "#8B4513", "#FC2727", "#FE7001", "#F6FF00", "#10FF00", "#015411", "#0771FF", "#241ED9", "#550068"]}
						title={"Backgroundcolor"}
						titleStyles={{fontSize: 30}}
					/>
				</View>
			)}
			<Button onPress={initializeDrawingPanel} title={buttonText}/>
			
			{hideOptions && (
				<ColorPalette
					onChange={changeColor}
					defaultColor={selectedColor}
					colors={["#FFFFFF", "#000000", "#C0C1BB", "#8B4513", "#FC2727", "#FE7001", "#F6FF00", "#10FF00", "#015411", "#0771FF", "#241ED9", "#550068"]}
					title={"Pixelcolor"}
					titleStyles={{fontSize: 30}}
				/>
			)}

			{hideOptions && (
				<Switch
					trackColor={{ false: "#767577", true: "#81b0ff" }}
					thumbColor={isEnabled ? "#f5dd4b" : "#f4f3f4"}
					ios_backgroundColor="#3e3e3e"
					onValueChange={toggleSwitch}
					value={isEnabled}
				/>
			)}

			<View style={editorStyles.wrapper}>
				<View style={editorStyles.drawingPanel}>
					{hideOptions && (
						<DrawingPanel 
							testID="drawingPanel"
							width={panelWidth}
							height={panelHeight}
							selectedColor = {selectedColor}
							newBackground = {newBackground}
						/>
					)}
				</View>
				<View style={editorStyles.grid}>
					{hideGrid && (
						<Grid 
							width={panelWidth}
							height={panelHeight}
							selectedColor = {selectedColor}
						/>
					)}
				</View>
			</View>
		</View>
		

	);
}


