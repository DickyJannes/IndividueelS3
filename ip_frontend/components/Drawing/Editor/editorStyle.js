import EStyleSheet from "react-native-extended-stylesheet";

const styles = EStyleSheet.create({
	editor: {
		backgroundColor: "#2b2929",
		borderColor: "red",
		borderWidth: 1,
		borderRadius: "0.25rem",
		paddingVertical: "1.5rem",
		paddingHorizontal: "4rem",
		marginTop: "2rem",
		marginBottom: "2rem",

		display: "flex",
		flexDirection: "column",
		alignItems: "center"
	},
	errorMessage: {
		fontSize: 16,
		color: "red"
	},
	h1: {
		display: "block",
		fontSize: "2rem",
		marginTop: "0.67rem",
		marginBottom: "1.5rem",
		marginLeft: 0,
		marginRight: 0,
		fontWeight: "bold"
	},
	h2: {
		display: "block",
		fontSize: "1.5rem",
		marginTop: "0.83rem",
		marginBottom: "1rem",
		marginLeft: 0,
		marginRight: 0,
		fontWeight: "bold"
	},
	options: {
		display: "flex",
		marginBottom: "3rem",
		justifyContent: "center",
		flexDirection: "row"
	},
	option: {
		display: "flex",
		flexDirection: "column",
		textAlign: "center"
	},
	panelInput: {
		height: "5rem",
		width: "5rem",
		fontSize: "2rem",
		textAlign: "center",
		backgroundColor: "#131313",
		color: "white",
		borderWidth: 1,
		borderColor: "grey",
		borderRadius: "0.25rem",
		marginTop: 0,
		marginRight: "1rem",
		marginLeft: "1rem",
		marginBottom: "0.5rem",
		paddingHorizontal: "1rem"
	},
	button: {
		fontSize: "1.25rem",
		backgroundColor: "red",
		color: "white",
		paddingVertical: "1rem",
		paddingHorizontal: "3rem",
		borderRadius: "0.25rem",
		borderWidth: 1,
		borderColor: "transparent",
		marginBottom: "2rem",
		underlayColor: "green"
	},
	circle_picker: {
		marginBottom: "1.5rem",
	},

	wrapper: {
		position: "relative"
	},
	drawingPanel: {
		zIndex: 1, // works on ios
		elevation: 1, // works on android
	},
	grid: {
		position: "absolute",
		top: 0,
		right: 0,
		zIndex: 0, // works on ios
		elevation: 0, // works on android
	}
});
  
export default styles;