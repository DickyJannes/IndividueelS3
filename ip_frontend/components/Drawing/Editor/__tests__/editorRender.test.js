import React from 'react';
import renderer from 'react-test-renderer';
import {render, fireEvent, waitFor} from '@testing-library/react-native';

import Editor from '../Editor';

test('renders editor correctly', () => {
  const tree = renderer.create(<Editor />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('buttonText should change when clicking on the drawingbutton', () => {
  const {getByTestId, getByText, queryByTestId, toJSON} = render(<Editor />)

  const button = getByText('start drawing')
  fireEvent.press(button)

  expect(getByText('reset')).not.toBeFalsy();
})

test('Drawingpanel starts as not visible but is visible after clicking button', async () => {
  const {getByTestId, getByText, queryByTestId, toJSON} = render(<Editor />)

  const button = getByText('start drawing')
  const drawingPanel = queryByTestId("drawingPanel")

  expect(drawingPanel).toBeNull();
  fireEvent.press(button)

  

})