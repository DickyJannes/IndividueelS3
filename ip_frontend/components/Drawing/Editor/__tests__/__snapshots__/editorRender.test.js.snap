// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders editor correctly 1`] = `
<View>
  <Text>
    Pixel Editor
  </Text>
  <Text>
    Enter Panel Dimensions
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <View>
      <View>
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          onChange={[Function]}
          rejectResponderTermination={true}
          testID="inputWidth"
          underlineColorAndroid="transparent"
          value={8}
        />
        <Text>
          Width
        </Text>
      </View>
      <View>
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          onChange={[Function]}
          rejectResponderTermination={true}
          underlineColorAndroid="transparent"
          value={8}
        />
        <Text>
          Height
        </Text>
      </View>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "black",
          },
          Object {
            "fontSize": 30,
          },
        ]
      }
    >
      Backgroundcolor
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "center",
          },
          Object {},
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      >
        <Text
          adjustsFontSizeToFit={true}
          style={
            Object {
              "color": "#000000",
              "fontSize": 20,
            }
          }
        >
          ✔︎
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#000000",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#C0C1BB",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#8B4513",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FC2727",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FE7001",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#F6FF00",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#10FF00",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#015411",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#0771FF",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#241ED9",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#550068",
            "borderColor": "rgba(0,0,0,0.2)",
            "borderRadius": 15,
            "borderWidth": 1,
            "elevation": 5,
            "height": 30,
            "justifyContent": "center",
            "marginHorizontal": 10,
            "marginVertical": 10,
            "opacity": 1,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 2,
            },
            "shadowOpacity": 0.25,
            "width": 30,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        start drawing
      </Text>
    </View>
  </View>
  <View>
    <View />
    <View />
  </View>
</View>
`;
